### PR TITLE
Add HarmonicSchwarzschild analytic solution

### PR DESCRIPTION
**Pull request by macedo22**
## Proposed changes

This PR adds `gr::Solutions::HarmonicSchwarzschild`: Schwarzschild in Cartesian coordinates that satisfies the harmonic gauge condition.

The equations implemented are referenced from [the SpEC implementation](https://github.com/sxs-collaboration/spec/blob/master/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/HarmonicSchwarzschild.cpp).

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.hpp
Comment (nilsvu) [inline src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.hpp]: (nitpick) prefer explicit types (`std::array<double, volume_dim>` here) unless they are particularly verbose

File: src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.hpp
Comment (nilsvu) [inline src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.hpp:23]: (nitpick) you're missing some includes, e.g. `TensorExpressions` in the cpp. Nothing to hold up this PR, but please do another pass over the includes when you get the chance. You can delete all the IWYU comments if you want (we don't actively use IWYU at the moment).

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/Test_HarmonicSchwarzschild.cpp
Comment (nilsvu) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/Test_HarmonicSchwarzschild.cpp:219]: Please add a test that the solution solves the Einstein equations. See `TestHelpers::VerifyGrSolution` in `tests/Unit/Helpers/PointwiseFunctions/AnalyticSolutions/TestHelpers.hpp`. For me that is the most important test. Re-implementations of the equations, as done in this function, have the tendency to have the same bugs as the original implementation ;)

**Review by nilsvu — CHANGES_REQUESTED** [review]
Looks great, I'm impressed by your level of documentation. Keep it up!

The only thing I'd like you to add is the test mentioned below.

**Comment by macedo22** [comment]
I am very grateful to @geoffrey4444 for going through the SpEC code and translating it into LaTeX! It made possible intermediates easy to spot and documentation quicker to write.

I'll push up the fixes and new test as soon as I can.

File: src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.hpp
Comment (macedo22) [reply src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.hpp:23]: I'll do another pass on the includes!

Regarding `TensorExpressions` specifically, though, mine and @nilsdeppe 's thought was to have `TensorExpressions` be considered part of the `Tensor` interface in that when you include `Tensor.hpp`, with it comes all of `TensorExpressions`. We currently just include all of the `TensorExpression` header files in `Tensor.hpp` to do this. This is one way of not having to have people include each individual TE file that they need.

Comment (nilsvu) [reply src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.hpp:23]: Sounds good, I believe Tensor.hpp is missing though

File: src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.hpp
Comment (nilsvu) [inline src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.hpp:39]: Reading through these docs I have a few more suggestions. Please pick which you find reasonable to include in this PR; we can also defer some to follow-ups:

- Could you add to these docs a few notes on what makes the solution "harmonic", ideally with a reference into a book or paper? For instance, you could refer to Sec. 4.3 in the Baumgarte/Shapiro book (or another book/paper of your choice!) and/or to https://journals.aps.org/prd/abstract/10.1103/PhysRevD.56.4775.
- (Edited) Furthermore, harmonic slicing is not unique, so it would be good to add here which harmonic coords are chosen. Even standard Schwarzschild coords are harmonic, but they are not horizon-penetrating. So I think here we mean harmonic + horizon-penetrating time slices, which are unique I think. They still leave a choice of spatial coords, which we could also make harmonic. Is your implementation here Eq. (45)+ in https://journals.aps.org/prd/abstract/10.1103/PhysRevD.56.4775 (I'm not sure if those are also spatially harmonic)? If so, that would be very useful information to add to these docs. However, I just looked at Eq. (47), which is `\beta^r = \alpha^2 (2M / r)^2`, which differs slightly from your Eq. (16): `\beta^r = \alpha^2 2M / (M + r)`, so maybe there's a radial transformation that I'm missing? @geoffrey4444 and @markscheel could you confirm / correct me here?
- I also suggest you add a test that verifies the harmonic coordinate conditions, e.g. a selection of Eqs. (4.42), (4.44), (4.45) in Baumgarte/Shapiro (whichever are easiest to implement).

Comment (kidder) [reply src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.hpp:39]: Standard Schwarzschild coordinates are not harmonic.  The simplest transformation to get a harmonic coordinate system from standard Schwarzschild coordinates is `r_H = r_S - M`.  Harmonic coordinates are also known as de Donder coordinates.  Goes back at least to a paper by Lanczos (1922).

Comment (geoffrey4444) [reply src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.hpp:39]: @macedo22 's implementation is just based on the implementation in SpEC, which @kidder did 16 years ago. Presumably it just used the transformation you mentioned. @kidder ? The code itself has no documentation explaining how the form of Schwarzschild was arrived at: https://github.com/sxs-collaboration/spec/blame/ad7fb23be8de1a725b38bef92c5a76de13524a97/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/HarmonicSchwarzschild.hpp

File: src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.hpp
Comment (geoffrey4444) [inline src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.hpp]: Just to confirm, the parameter x passed into these functions is x^i - c^I, not x^i? I would have thought these functions just take x^I, but this line in the documentation makes me think otherwise. Does Kerr Schild work the same way?

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/Test_HarmonicSchwarzschild.cpp
Comment (geoffrey4444) [reply tests/Unit/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/Test_HarmonicSchwarzschild.cpp:219]: I second this suggestion

File: src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.cpp
Comment (geoffrey4444) [inline src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.cpp:581]: I checked that all of the equations implemented here are correct in the sense that they agree with my understanding of what the SpEC code does and with the documentation given here, except I never worked out the inverse metric explicitly. Maybe add a quick test to the unit test that confirms metric * inverse metric == identity?

**Review by geoffrey4444 — CHANGES_REQUESTED** [review]

File: src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.hpp
Comment (macedo22) [reply src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.hpp]: Whoops, I think this is a documentation mistake on my part. It  should just be x^i, not X^i = x^i - C^i. I'll fix the documentation

File: src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.hpp
Comment (macedo22) [reply src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.hpp:39]: @nilsvu @geoffrey4444 I've gone ahead and addressed almost all the change requests in the fixup I just pushed. The remaining tasks: documenting a literature reference (not sure what is best to cite/reference, please let me know!) and getting all of the new harmonic condition tests to pass (more on this below).

Regarding testing that the implementation satisfies eq 4.42, 4.44, and 4.45 (Baumgarte & Shapiro), it seems that only 4.44 holds (I've marked 4.42 and 4.45 with `TODO` comments in the test file in the fixup). What's strange is that I also copied over the SpEC implementation (and tweaked syntax so it compiles), and this PR's implementation matches the previous SpEC implementation's results in an additional test case I included in the fixup, so I'm not sure why 4.42 and 4.45 are not also satisfied.

Am I perhaps setting up the test in a weird way (e.g. weird center / bounds / coords / mass)?

Comment (nilsvu) [reply src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.hpp:39]: Thanks for checking so thoroughly. I believe what we have here are only _time-harmonic_ slices of Schwarzschild, Eq. 4.44. The spatial coords are not harmonic, Eq. 4.45. You actually copied a comment from SpEC in the test below that confirms this. So I suggest you add this information to the docs and remove the failing checks for Eq. 4.42 and 4.45.

Comment (macedo22) [reply src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.hpp:39]: Thank you!

**Comment by macedo22** [comment]
@nilsvu @geoffrey4444 All comments have been addressed outside of citing literature, so I think the PR is ready for re-review. If there is a specific resource to cite, let me know and I can add it!

**Comment by nilsvu** [comment]
Please just add to the docs what you already know, i.e. refer to Eq. (4.44) in Baumgarte/Shapiro. Let's also check how this solution relates to the one listed in the Cook/Scheel paper that I linked above. @markscheel do you know? (I'm on mobile now, can do the math later)

File: src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.hpp
Comment (nilsvu) [reply src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.hpp:39]: With your recent corrections (missing square) it looks like your implementation differs from Eq. (45)--(50) in https://journals.aps.org/prd/abstract/10.1103/PhysRevD.56.4775 only by `r -> M + r` (I haven't checked this thoroughly). @geoffrey4444 and @markscheel what do you make of that?
- Should this class implement Eq. (45)--(50) exactly instead?
- Any ideas why SpEC has the additional `+M`?
- Is my assumption correct that Eq. (45)--(50) are the _time and space_ harmonic Kerr slices presented earlier in the paper, in the limit of zero spin? (I don't see trivially how Eq. (22) becomes Eq. (45), but I haven't gone through the math.)

@macedo22  Please add the information we gather here to the docs.

**Comment by nilsdeppe** [comment]
@nilsvu @geoffrey4444 What's needed to keep this moving?

**Comment by nilsvu** [comment]
I'd like to answer the questions in the thread above: https://github.com/sxs-collaboration/spectre/pull/3829#discussion_r820247536

File: src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.hpp
Comment (nilsvu) [reply src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.hpp:39]: @macedo22: @markscheel and I discussed this and looked at the paper + the SpEC implementation. Our hypothesis is this:

- SpEC has a `TimeHarmonicSchwarzschild` solution that is _only_ time harmonic. This is Eq. (45)--(50) in the 1997 paper.
- What SpEC calls `HarmonicSchwarzschild`, and what you have implemented, is time _and_ spatially harmonic. It differs from `TimeHarmonicSchwarzschild`, and hence from the paper, by `r -> M + r`. It is the zero-spin limit of the `HarmonicKerr` solution in SpEC.

We want time _and_ spatially harmonic coordinates (so Lorentz boosts preserve the harmonic time slicing). If our assumptions above are correct, then this is what you have implemented and everything is fine. Our suggestion is therefore:

1. Please add this information to the docs.
2. Please check again that the spatial coordinates are also harmonic, i.e., Eq. 4.45 in Baumgarte/Shapiro (the one that you said failed earlier). If this still fails then there's something we don't understand..

Comment (macedo22) [reply src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.hpp:39]: I checked again and as before, eq 4.42 and 4.45 fail even though the unit tests shows that this spectre implementation matches the spec imlpementation result. It's definitely possible I have made a mistake somewhere, so I will keep looking into it, but I just wanted to let you know these conditions fail even after having already looked at it for a moderate amount of time and adding things like the spec impl test

Comment (macedo22) [reply src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.hpp:39]: Yep, just fixed a mistake I made for 4.42 (I made a bad assumption about the equation)

Comment (macedo22) [reply src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.hpp:39]: @nilsvu @geoffrey4444 In the most recent fixup (0fa1e008d), I fixed my test for 4.42 and got it to pass. However, 4.45 still won't pass. I've tried the following things:
- Added the spec implementation and tested against the spectre implementation --> the unit test shows they are equivalent
- Rewrote the 4.45 equation in python to make sure it wasn't a bug in `TensorExpression`s ---> same failing result
- Reimplemented the whole Harmonic Schwarzschild solution in Mathematica and checked 4.45 as well ---> nearly the same failing result (some numerical error probably)
- Instead of calculating the derivative of the natural log of the lapse analytically as I had previously done, the most recent fixup uses a numerical derivative instead to make sure the condition failing wasn't due to me making a mistake in the analytical expression I came up with on paper ---> using the numerical derivative didn't fix the issue

After these attempts, I'm not sure what else I can try, so I was wondering if maybe it would be good for a second pair of eyes to take a close look at my unit test and make sure I'm not doing something weird, like setting it up with strange parameters, or misinterpreting eq 4.45. If not, let me know what else I can try!

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/Test_HarmonicSchwarzschild.cpp
Comment (nilsvu) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/Test_HarmonicSchwarzschild.cpp]: With this numerical derivative the `CHECK_ITERABLE_APPROX` below likely fails due to discretization error on the grid. Write the term `\partial_j \ln\alpha` as `\frac{\partial_j \alpha}{\alpha}` and use the analytic expressions. Then you don't need to set up a grid for taking derivatives in this test anymore.

File: src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.hpp
Comment (nilsvu) [reply src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.hpp:39]: That you got Eq. 4.42 means that the coordinates are indeed harmonic in space and time, so that's good! It also means that both Eq. 4.44 and 4.45 should pass, since they are just the 3+1 version of Eq. 4.42.

In my comment below I suggest you get rid of the numerical derivative, but from your comment here I understand that you have tried that already (still do it please). So I went ahead and derived Eq. 4.45 from Eq. 4.42 and got almost the result printed in Baumgarte/Shapiro, except for a minus sign. That might well be my mistake, but it might also be a typo in the book, so please check:

```tex
% Eq. 4.45 (check sign of \Gamma contraction term)
(\partial_t - \beta^j \partial_j) \beta^i
  = -\alpha^2 (\gamma^{ij} \frac{\partial_j \alpha}{\alpha} - \gamma^{jk} \Gamma^i_{jk})
```

You can also squash all fixup commits together to clean up this PR a bit, and add the information we gathered here to the docs.

Comment (macedo22) [reply src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.hpp:39]: Oh my gosh @nilsvu the book did just have the sign incorrect. Switching it makes the test pass! I was starting to feel crazy. Thanks so much! I'll update things as soon as I can

Comment (nilsvu) [reply src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.hpp:39]: Nice!

Comment (macedo22) [reply src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.hpp:39]: @nilsvu I've gone ahead and squashed the previous fixups, rebased, and added two new fixups so that you can see the new changes:
- first fixup: updates the test file to use the anlytical derivative and fixes the 4.45 check
- second fixup: documenting the result of our discussion on the equations in this thread (unsure if I described this correctly)

Comment (macedo22) [reply src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.hpp:39]: 3rd fixup: removing unneeded includes after removing numerical derivative

File: src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.hpp
Comment (nilsvu) [inline src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.hpp]: I suggest something like this:

> This solution represents a Schwarzschild black hole in coordinates that are harmonic in both space and time, as well as horizon-penetrating. Therefore, this solution fulfills the harmonic coordinate conditions, Eq. (4.42) to (4.45) in \cite BaumgarteShapiro :
>
> [write equations 4.42, 4.44, and 4.45 here]
>
> (Note that Eq. 4.45 in \cite Baumgarte is missing a minus sign in front of the $\Gamma$-contraction term)
>
> We implement Eqs. (45)--(50) in \cite Cook1997qc , which represent the zero-spin limit of the time-harmonic and horizon-penetrating slices of Kerr spacetime presented in the paper. We add the radial transformation `r -> r + M` to make the spatial coordinates harmonic as well (see Eq. (43) in \cite Cook1997qc ), so the coordinates remain harmonic under boosts.

File: src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.cpp
Comment (nilsvu) [inline src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/HarmonicSchwarzschild.cpp]: delete this line, the `NOLINT`, and the `std::move`. Make the `center` argument a `const &`.

**Review by nilsvu — DISMISSED** [review]
Looks great, go ahead and squash this in, then let's get this merged.

**Comment by macedo22** [comment]
@nilsvu Ready for re-review

**Review by nilsvu — APPROVED** [review]

**Comment by nilsvu** [comment]
@geoffrey4444 please check if your comments have been addressed.

**Review by geoffrey4444 — APPROVED** [review]

**Pull request merged**
